drivers: spi: Added ZynqMP Generic Quad SPI driver

robhancocksed opened the pull request:
Add a driver for the Generic Quad SPI hardware that is part of the Xilinx MPSoC. This is mostly commonly used with QSPI flash devices which are also used for initial image loading, but can also be used with other SPI devices.

Also added corresponding settings for the QSPI device and emulated flash devices for the QEMU Cortex-R5 board.

robhancocksed committed:
drivers: spi: Added ZynqMP Generic Quad SPI driver

Add a driver for the Generic Quad SPI hardware that is part of the
Xilinx MPSoC. This is mostly commonly used with QSPI flash devices which
are also used for initial image loading, but can also be used with other
SPI devices.

Signed-off-by: Robert Hancock <robert.hancock@calian.com>

robhancocksed committed:
boards: qemu: cortex_r5: Added QSPI settings

Added device tree entries for the QSPI device and the flash devices
which are present behind it on this emulated board.

Signed-off-by: Robert Hancock <robert.hancock@calian.com>

robhancocksed commented:
Ping.. any feedback?

robhancocksed commented:
@AmitMahapatra , note that if you are testing on boards with certain Micron flash devices, such as ZCU102, you will need the patches from https://github.com/zephyrproject-rtos/zephyr/pull/88467 since those devices require the flag status register to be read after each program operation, otherwise they get into a state where all following reads return zeros. You will also need the appropriate flags in DT to enable that usage. These are the DT entries I have at the board level which are working for me on ZCU102 with those patches in place:

```
&qspi {
	status = "okay";
	clock-frequency = <124987511>;

	flash@0 {
		compatible = "jedec,spi-nor";
		#address-cells = <1>;
		#size-cells = <1>;
		reg = <0x0>;
		spi-max-frequency = <166000000>;
		use-flag-status-register;
		use-fast-read;
	};

	flash@1 {
		compatible = "jedec,spi-nor";
		#address-cells = <1>;
		#size-cells = <1>;
		reg = <0x1>;
		spi-max-frequency = <166000000>;
		use-flag-status-register;
		use-fast-read;
	};
};

```

robhancocksed commented on drivers/spi/spi_xlnx_zynqmp_gqspi.c:
```diff
@@ -0,0 +1,662 @@
+		return -ENOTSUP;
+	}
+
+	if ((spi_cfg->operation & SPI_LINES_MASK) == SPI_LINES_OCTAL) {
```
It does have support for basic dual/quad reads/writes which might be useful for some use cases. However, as you noted the SPI API would probably need more fine-grained control to be usable for QSPI flash in those modes, for example..

robhancocksed commented:
Rebased to main, and added a fix for the shared-data-bus DT parameter not taking effect properly.

robhancocksed committed:
dts: arm: xilinx: zynqmp: mark flash0 node as disabled

Currently ZynqMP board setups do not appear to be using the flash0 node
which is used for running in XIP mode from QSPI flash storage. Mark this
node as disabled to avoid tripping up the
drivers.flash.common.disable_spi_nor tests, which would otherwise try to
build with FLASH enabled and SPI_NOR disabled.

Signed-off-by: Robert Hancock <robert.hancock@calian.com>

robhancocksed commented:
Ping? Still looking for feedback.

robhancocksed commented:
Rebased to fix a conflict with other recently merged changes in zynqmp.dtsi.

michalsimek commented:
@henrikbrixandersen / @ibirnbaum can you please reivew it? We can't see any problem with this driver on our platforms.

tbursztyka commented on drivers/spi/spi_xlnx_zynqmp_gqspi.c:
```diff
@@ -0,0 +1,663 @@
+ * TODO: DMA mode not yet implemented
+ */
+
+enum ZynqMPGQSPIRegisters {
```
no camel case

robhancocksed replied:
Changed this.

tbursztyka commented on drivers/spi/spi_xlnx_zynqmp_gqspi.c:
```diff
@@ -0,0 +1,663 @@
+		bool complete = xlnx_zynqmp_gqspi_service_fifos(dev);
+		uint32_t wait_events = 0;
+
+		if (complete || async) {
```
async has nothing to do with being isr-based or not.

tbursztyka reviewed (changes requested):
There is a confusion about the async mode, which is only meant to do a cb based transceive. Internally the driver should do things interrupt based anyway.

tbursztyka commented on drivers/spi/spi_xlnx_zynqmp_gqspi.c:
```diff
@@ -0,0 +1,663 @@
+		 * For async mode, we need to read the RX FIFO and refill the TX FIFO
+		 * if needed here.
+		 * For sync mode, we do this in the caller's context to avoid doing too much
+		 * work in the ISR, so just post the event.
```
It's indeed an issue that we are missing a proper generic 2-halves irq handling subsystem in zephyr, for the controllers that have demanding irq work. The problem with your approach is you rely on caller's thread priority, which is not ideal.

Well, apply my comments on camel-case, and let's move on.

ibirnbaum commented:
> Looks like a test failure in sample.llext.modules.module_build for the qemu_cortex_r5/zynqmp_rpu board. I assume that is being run as this driver is enabled in that build, but the failure (prefetch abort/permission fault) seems to have nothing to do with it.

Without having seen the error message, I could imagine that this is fallout from the recent PR that enables the MPU in the R5 by default (although there were no build or run errors in that PR's CI runs, and qemu_cortex_r5 was definitely included). Using LLEXT to load further code requires MPU re-configuration as all RAM outside the original text section is non-executable, hence the prefetch abort. I'm just wondering why this didn't show up in the CI when this change was made. I'll look into that over the weekend. This is likely not down to your driver as you're covered by the 'peripherals' region in the MPU region table.

tbursztyka commented:
Could you rebase to get CI passing?

robhancocksed commented:
Not stale, but needs some input from @ibirnbaum on how to resolve the LLEXT test failure on ZynqMP.

robhancocksed commented:
Not sure what is going on with the CI test failures. One job failed in the Check Environment phase, and the other has some shell command tests which are failing with timeouts. But those builds don't even have this driver enabled in the config, so it doesn't seem related to anything changed here?

maass-hamburg commented:
reopened to re-trigger ci

robhancocksed commented:
Rebased to current main and fixed a merge conflict as well as dropped a patch for an issue that was already fixed another way. However, still getting some test failures that appear to have nothing to do with the changes made. Frustrating..

tbursztyka commented:
CI is failing on non relevant tests it seems? rebase and push again, let's see.

robhancocksed commented:
Rebased to main and added an include for zephyr/sys/minmax.h that's now needed.

robhancocksed commented:
@tbursztyka Looks like CI is clean now finally..

nashif commented:
@maass-hamburg  can you please revisit?
